In [1]:
# import 'Pandas'
# It is utilised to format the data structure. It means it can convert any data structure into pandas data frame.
import pandas as pd

# import 'numpy'
# It is used to find statistical summary.
import numpy as np

# import subpackages of Matplotlib
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap

# import 'seaborn'
import seaborn as sns

In [2]:
# Load the data and name it df_houseprice
# Load the csv file
# store the data in 'df_houseprice'
df_houseprice = pd.read_csv("House_Price1.csv")

# display first five observation using head()
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200      None        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

## Data Overview

In [3]:
# Name the Columns 
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

**Interpretation**

 - `price`:Price of the House.
 - `resid_area`:Residential Area where my house is located.
 - `air_qual`:Air quality of the atmosphere.
 - `room_num`:The plot no.of the House.
 - `age`:The age of the people buying the house.
 - `distance`:The distance of the school

In [4]:
# Basic Info of the Data Set
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    506 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


**Interpretation**

 - we have 13 numerical column and 3 categorial column
 - we have null values in the data
 - we have total 16 columns in the data

In [5]:
# Shape of the Data
df_houseprice.shape

(506, 16)

**Interpretation**
 - We have total 16 columns.
 - We have 506 observations.

In [6]:
# Summarize the basic statistics of Data
df_houseprice.describe().T

count       mean        std        min        25%        50%  \
price        506.0  22.528854   9.182176   5.000000  17.025000  21.200000   
resid_area   506.0  41.136779   6.860353  30.460000  35.190000  39.690000   
air_qual     506.0   0.554695   0.115878   0.385000   0.449000   0.538000   
room_num     506.0   6.284634   0.702617   3.561000   5.885500   6.208500   
age          506.0  68.574901  28.148861   2.900000  45.025000  77.500000   
distance     506.0   3.971996   2.108532   1.130000   2.270000   3.385000   
teachers     506.0  21.544466   2.164946  18.000000  19.800000  20.950000   
poor_prop    506.0  12.653063   7.141062   1.730000   6.950000  11.360000   
n_hos_beds   498.0   7.899767   1.476683   5.268000   6.634500   7.999000   
n_hot_rooms  506.0  13.041605   5.238957  10.057600  11.189800  12.720000   
rainfall     506.0  39.181818  12.513697   3.000000  28.000000  39.000000   
parks        506.0   0.054454   0.010632   0.033292   0.046464   0.053507   
Sold         506.0   0.454545   0.498422   0.000000   0.000000   0.000000   

                   75%         max  
price        25.000000   50.000000  
resid_area   48.100000   57.740000  
air_qual      0.624000    0.871000  
room_num      6.623500    8.780000  
age          94.075000  100.000000  
distance      5.367500   12.320000  
teachers     22.600000   27.400000  
poor_prop    16.955000   37.970000  
n_hos_beds    9.088000   10.876000  
n_hot_rooms  14.170800  101.120000  
rainfall     50.000000   60.000000  
parks         0.061397    0.086711  
Sold          1.000000    1.000000

**Interpretation**
 - We have outliers in the data, most of the outliers are in age column.

## Data Preprocessing

### a. Data Cleaning

In [7]:
# Show the first five observations
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200      None        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [8]:
# Rename the Columns
# Step 1 : Show the column names 
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [9]:
# Step 2 : Allocate new column names to old column names
# Make the dictionary{'Old column name':'New column name'}
d1={'price':'Price',
    'resid_area':'Residential_Area',
    'air_qual':'Air_Quality',
    'room_num':'Room_Numbers',
    'age':'Age',
    'distance':'Distance',
    'teachers':'Teachers',
    'poor_prop':'Proportion_Of_Poor_Population',
    'airport':'Airport',
    'n_hos_beds':'Number_Of_Hospital_Beds',
    'n_hot_rooms':'Number_Of_Hotel_Rooms',
    'waterbody':'Waterbody',
    'rainfall':'Rainfall',
    'bus_ter':'Bus_Terminal',
    'parks':'Parks',
    'Sold':'Sold'}

In [10]:
df_houseprice = df_houseprice.rename(columns=d1)

In [11]:
# Authenticate the Results
df_houseprice.columns

Index(['Price', 'Residential_Area', 'Air_Quality', 'Room_Numbers', 'Age',
       'Distance', 'Teachers', 'Proportion_Of_Poor_Population', 'Airport',
       'Number_Of_Hospital_Beds', 'Number_Of_Hotel_Rooms', 'Waterbody',
       'Rainfall', 'Bus_Terminal', 'Parks', 'Sold'],
      dtype='object')

### b. Null Value Handling

In [12]:
# Find the valid Null Values
df_houseprice.isnull().sum()

Price                            0
Residential_Area                 0
Air_Quality                      0
Room_Numbers                     0
Age                              0
Distance                         0
Teachers                         0
Proportion_Of_Poor_Population    0
Airport                          0
Number_Of_Hospital_Beds          8
Number_Of_Hotel_Rooms            0
Waterbody                        0
Rainfall                         0
Bus_Terminal                     0
Parks                            0
Sold                             0
dtype: int64

In [13]:
# Percentage missing value
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                            0.000
Residential_Area                 0.000
Air_Quality                      0.000
Room_Numbers                     0.000
Age                              0.000
Distance                         0.000
Teachers                         0.000
Proportion_Of_Poor_Population    0.000
Airport                          0.000
Number_Of_Hospital_Beds          1.581
Number_Of_Hotel_Rooms            0.000
Waterbody                        0.000
Rainfall                         0.000
Bus_Terminal                     0.000
Parks                            0.000
Sold                             0.000
dtype: float64

In [14]:
# Find the nature of the column
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Price                          506 non-null    float64
 1   Residential_Area               506 non-null    float64
 2   Air_Quality                    506 non-null    float64
 3   Room_Numbers                   506 non-null    float64
 4   Age                            506 non-null    float64
 5   Distance                       506 non-null    float64
 6   Teachers                       506 non-null    float64
 7   Proportion_Of_Poor_Population  506 non-null    float64
 8   Airport                        506 non-null    object 
 9   Number_Of_Hospital_Beds        498 non-null    float64
 10  Number_Of_Hotel_Rooms          506 non-null    float64
 11  Waterbody                      506 non-null    object 
 12  Rainfall                       506 non-null    int

In [15]:
# This is a categorical column so we can replace the null with mode.
df_houseprice.Waterbody.mode()

0    River
Name: Waterbody, dtype: object

In [16]:
# Replace the null in that column
df_houseprice.Waterbody = df_houseprice.Waterbody.fillna('River')

In [17]:
# Percentage missing value
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                            0.000
Residential_Area                 0.000
Air_Quality                      0.000
Room_Numbers                     0.000
Age                              0.000
Distance                         0.000
Teachers                         0.000
Proportion_Of_Poor_Population    0.000
Airport                          0.000
Number_Of_Hospital_Beds          1.581
Number_Of_Hotel_Rooms            0.000
Waterbody                        0.000
Rainfall                         0.000
Bus_Terminal                     0.000
Parks                            0.000
Sold                             0.000
dtype: float64

In [18]:
# In Number_Of_Hospital_Beds contain less than 10% values so we can drop null values
df_houseprice.dropna(inplace=True)

In [19]:
# Percentage missing value
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                            0.0
Residential_Area                 0.0
Air_Quality                      0.0
Room_Numbers                     0.0
Age                              0.0
Distance                         0.0
Teachers                         0.0
Proportion_Of_Poor_Population    0.0
Airport                          0.0
Number_Of_Hospital_Beds          0.0
Number_Of_Hotel_Rooms            0.0
Waterbody                        0.0
Rainfall                         0.0
Bus_Terminal                     0.0
Parks                            0.0
Sold                             0.0
dtype: float64

## Prepare Data for Machine Learning

In [20]:
# Seperate the target variable and independent variable
target = df_houseprice['Sold']

In [21]:
# Independent Variable
df_independent = df_houseprice.drop('Sold', axis = 1)
df_independent

Price  Residential_Area  Air_Quality  Room_Numbers   Age  Distance  \
0     24.0             32.31        0.538         6.575  65.2      4.35   
1     21.6             37.07        0.469         6.421  78.9      4.99   
2     34.7             37.07        0.469         7.185  61.1      5.03   
3     33.4             32.18        0.458         6.998  45.8      6.21   
4     36.2             32.18        0.458         7.147  54.2      6.16   
..     ...               ...          ...           ...   ...       ...   
501   22.4             41.93        0.573         6.593  69.1      2.64   
502   20.6             41.93        0.573         6.120  76.7      2.44   
503   23.9             41.93        0.573         6.976  91.0      2.34   
504   22.0             41.93        0.573         6.794  89.3      2.54   
505   19.0             41.93        0.573         6.030  80.8      2.72   

     Teachers  Proportion_Of_Poor_Population Airport  Number_Of_Hospital_Beds  \
0        24.7                           4.98     YES                    5.480   
1        22.2                           9.14      NO                    7.332   
2        22.2                           4.03      NO                    7.394   
3        21.3                           2.94     YES                    9.268   
4        21.3                           5.33      NO                    8.824   
..        ...                            ...     ...                      ...   
501      19.0                           9.67      NO                    9.348   
502      19.0                           9.08     YES                    6.612   
503      19.0                           5.64      NO                    5.478   
504      19.0                           6.48     YES                    7.940   
505      19.0                           7.88     YES                   10.280   

     Number_Of_Hotel_Rooms       Waterbody  Rainfall Bus_Terminal     Parks  
0                  11.1920           River        23          YES  0.049347  
1                  12.1728            Lake        42          YES  0.046146  
2                 101.1200            None        38          YES  0.045764  
3                  11.2672            Lake        45          YES  0.047151  
4                  11.2896            Lake        55          YES  0.039474  
..                     ...             ...       ...          ...       ...  
501                12.1792  Lake and River        27          YES  0.056006  
502                13.1648  Lake and River        20          YES  0.059903  
503                12.1912            None        31          YES  0.057572  
504                15.1760            None        47          YES  0.060694  
505                10.1520            None        45          YES  0.060336  

[498 rows x 15 columns]

## Seperate the Categorical Column from Independent Variable

In [22]:
df_cat = df_independent.select_dtypes(include = "object")
df_cat.head()

Airport Waterbody Bus_Terminal
0     YES     River          YES
1      NO      Lake          YES
2      NO      None          YES
3     YES      Lake          YES
4      NO      Lake          YES

In [23]:
# Encode the Categorical variable
encoded = pd.get_dummies(df_cat)
encoded = encoded.reset_index()

In [24]:
encoded.head()

index  Airport_NO  Airport_YES  Waterbody_Lake  Waterbody_Lake and River  \
0      0           0            1               0                         0   
1      1           1            0               1                         0   
2      2           1            0               0                         0   
3      3           0            1               1                         0   
4      4           1            0               1                         0   

   Waterbody_None  Waterbody_River  Bus_Terminal_YES  
0               0                1                 1  
1               0                0                 1  
2               1                0                 1  
3               0                0                 1  
4               0                0                 1

## Scale the Numerical Data

In [25]:
df_num = df_independent.select_dtypes(include = 'number')
df_num.head()

Price  Residential_Area  Air_Quality  Room_Numbers   Age  Distance  \
0   24.0             32.31        0.538         6.575  65.2      4.35   
1   21.6             37.07        0.469         6.421  78.9      4.99   
2   34.7             37.07        0.469         7.185  61.1      5.03   
3   33.4             32.18        0.458         6.998  45.8      6.21   
4   36.2             32.18        0.458         7.147  54.2      6.16   

   Teachers  Proportion_Of_Poor_Population  Number_Of_Hospital_Beds  \
0      24.7                           4.98                    5.480   
1      22.2                           9.14                    7.332   
2      22.2                           4.03                    7.394   
3      21.3                           2.94                    9.268   
4      21.3                           5.33                    8.824   

   Number_Of_Hotel_Rooms  Rainfall     Parks  
0                11.1920        23  0.049347  
1                12.1728        42  0.046146  
2               101.1200        38  0.045764  
3                11.2672        45  0.047151  
4                11.2896        55  0.039474

In [26]:
from sklearn.preprocessing import StandardScaler

In [27]:
# Initiate the Standard Scaler
ss = StandardScaler()

In [28]:
# Transform The Data 
scale = ss.fit_transform(df_num)
scale

array([[ 0.15479189, -1.28244435, -0.13709359, ..., -0.35211111,
        -1.30206786, -0.47313205],
       [-0.10659165, -0.59005882, -0.73319492, ..., -0.16601943,
         0.21519397, -0.77522463],
       [ 1.32012686, -0.59005882, -0.73319492, ..., 16.71034024,
        -0.10422957, -0.81123666],
       ...,
       [ 0.14390091,  0.11687263,  0.16527665, ..., -0.16252832,
        -0.66322077,  0.30293265],
       [-0.06302773,  0.11687263,  0.16527665, ...,  0.40379146,
         0.6144734 ,  0.59750894],
       [-0.38975716,  0.11687263,  0.16527665, ..., -0.54943507,
         0.45476163,  0.56369225]])

In [30]:
df_scale = pd.DataFrame(scale, columns = df_num.columns)
df_scale = df_scale.reset_index()
df_scale

index     Price  Residential_Area  Air_Quality  Room_Numbers       Age  \
0        0  0.154792         -1.282444    -0.137094      0.408975 -0.115644   
1        1 -0.106592         -0.590059    -0.733195      0.189974  0.369741   
2        2  1.320127         -0.590059    -0.733195      1.276446 -0.260905   
3        3  1.178544         -1.301354    -0.828226      1.010516 -0.802977   
4        4  1.483492         -1.301354    -0.828226      1.222407 -0.505369   
..     ...       ...               ...          ...           ...       ...   
493    493 -0.019464          0.116873     0.165277      0.434572  0.022531   
494    494 -0.215501          0.116873     0.165277     -0.238073  0.291796   
495    495  0.143901          0.116873     0.165277      0.979230  0.798438   
496    496 -0.063028          0.116873     0.165277      0.720411  0.738208   
497    497 -0.389757          0.116873     0.165277     -0.366061  0.437057   

     Distance  Teachers  Proportion_Of_Poor_Population  \
0    0.172845  1.462167                      -1.066368   
1    0.476100  0.304963                      -0.484136   
2    0.495053  0.304963                      -1.199330   
3    1.054179 -0.111631                      -1.351886   
4    1.030487 -0.111631                      -1.017382   
..        ...       ...                            ...   
493 -0.637413 -1.176258                      -0.409958   
494 -0.732180 -1.176258                      -0.492534   
495 -0.779564 -1.176258                      -0.973995   
496 -0.684797 -1.176258                      -0.856429   
497 -0.599506 -1.176258                      -0.660486   

     Number_Of_Hospital_Beds  Number_Of_Hotel_Rooms  Rainfall     Parks  
0                  -1.640299              -0.352111 -1.302068 -0.473132  
1                  -0.384875              -0.166019  0.215194 -0.775225  
2                  -0.342847              16.710340 -0.104230 -0.811237  
3                   0.927490              -0.337843  0.454762 -0.680402  
4                   0.626514              -0.333593  1.253320 -1.404723  
..                       ...                    ...       ...       ...  
493                 0.981720              -0.164805 -0.982644  0.155116  
494                -0.872945               0.022197 -1.541636  0.522886  
495                -1.641654              -0.162528 -0.663221  0.302933  
496                 0.027273               0.403791  0.614473  0.597509  
497                 1.613499              -0.549435  0.454762  0.563692  

[498 rows x 13 columns]

In [35]:
X = pd.concat([df_scale,encoded],axis = 1)
X.head()

index     Price  Residential_Area  Air_Quality  Room_Numbers       Age  \
0      0  0.154792         -1.282444    -0.137094      0.408975 -0.115644   
1      1 -0.106592         -0.590059    -0.733195      0.189974  0.369741   
2      2  1.320127         -0.590059    -0.733195      1.276446 -0.260905   
3      3  1.178544         -1.301354    -0.828226      1.010516 -0.802977   
4      4  1.483492         -1.301354    -0.828226      1.222407 -0.505369   

   Distance  Teachers  Proportion_Of_Poor_Population  Number_Of_Hospital_Beds  \
0  0.172845  1.462167                      -1.066368                -1.640299   
1  0.476100  0.304963                      -0.484136                -0.384875   
2  0.495053  0.304963                      -1.199330                -0.342847   
3  1.054179 -0.111631                      -1.351886                 0.927490   
4  1.030487 -0.111631                      -1.017382                 0.626514   

   ...  Rainfall     Parks  index  Airport_NO  Airport_YES  Waterbody_Lake  \
0  ... -1.302068 -0.473132      0           0            1               0   
1  ...  0.215194 -0.775225      1           1            0               1   
2  ... -0.104230 -0.811237      2           1            0               0   
3  ...  0.454762 -0.680402      3           0            1               1   
4  ...  1.253320 -1.404723      4           1            0               1   

   Waterbody_Lake and River  Waterbody_None  Waterbody_River  Bus_Terminal_YES  
0                         0               0                1                 1  
1                         0               0                0                 1  
2                         0               1                0                 1  
3                         0               0                0                 1  
4                         0               0                0                 1  

[5 rows x 21 columns]

In [36]:
# Split the Dataset For Training and Testing

import statsmodels
import statsmodels.api as sm

X = sm.add_constant(X)
X.head()

index     Price  Residential_Area  Air_Quality  Room_Numbers       Age  \
0      0  0.154792         -1.282444    -0.137094      0.408975 -0.115644   
1      1 -0.106592         -0.590059    -0.733195      0.189974  0.369741   
2      2  1.320127         -0.590059    -0.733195      1.276446 -0.260905   
3      3  1.178544         -1.301354    -0.828226      1.010516 -0.802977   
4      4  1.483492         -1.301354    -0.828226      1.222407 -0.505369   

   Distance  Teachers  Proportion_Of_Poor_Population  Number_Of_Hospital_Beds  \
0  0.172845  1.462167                      -1.066368                -1.640299   
1  0.476100  0.304963                      -0.484136                -0.384875   
2  0.495053  0.304963                      -1.199330                -0.342847   
3  1.054179 -0.111631                      -1.351886                 0.927490   
4  1.030487 -0.111631                      -1.017382                 0.626514   

   ...  Rainfall     Parks  index  Airport_NO  Airport_YES  Waterbody_Lake  \
0  ... -1.302068 -0.473132      0           0            1               0   
1  ...  0.215194 -0.775225      1           1            0               1   
2  ... -0.104230 -0.811237      2           1            0               0   
3  ...  0.454762 -0.680402      3           0            1               1   
4  ...  1.253320 -1.404723      4           1            0               1   

   Waterbody_Lake and River  Waterbody_None  Waterbody_River  Bus_Terminal_YES  
0                         0               0                1                 1  
1                         0               0                0                 1  
2                         0               1                0                 1  
3                         0               0                0                 1  
4                         0               0                0                 1  

[5 rows x 21 columns]

In [37]:
# Splitting in train and test subsets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, target, random_state = 10, test_size = 0.2)

In [38]:
# print dimensios of train set
print('X_train', X_train.shape)
print('y_train', y_train.shape)

# print dimensions of test set
print('X_test', X_test.shape)
print('y_test', y_test.shape)

X_train (398, 21)
y_train (398,)
X_test (100, 21)
y_test (100,)


In [40]:
# Decision Tree Classifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Create a Decision Tree Classifier
tree_model = DecisionTreeClassifier()

# Fit the model to the Training Data
tree_model.fit(X_train,y_train)

# Predict the Training Data
predictions = tree_model.predict(X_test)

In [41]:
# Check the Accuracy
accuracy = accuracy_score(y_test, predictions)
print("Accuracy :", accuracy)

Accuracy : 0.61


In [ ]:
# 